src/file_pipe_log/file_mgr.rs

```diff
@@ -0,0 +1,751 @@
+        )
+    }
+}
+
+/// A collection of files for managing all log files.
+///
+/// Log files consist of two parts:
+/// - `Stale` part, obsolete files but temporarily stashed for recycling.
+/// - `Active` part, active files for accessing.
+pub struct FileCollection<F: FileSystem> {
+    file_system: Arc<F>,
+    queue: LogQueue,
+    paths: Paths,
+    file_format: LogFileFormat,
+    target_file_size: usize,
+    capacity: usize,
+    /// File list to collect all active files
+    active_files: CachePadded<RwLock<FileList<F>>>,
+    /// File list to collect all stale files
+    stale_files: CachePadded<RwLock<FileList<F>>>,
+    /// Flag remarking whether current collection is initialized
+    /// from an empty active file list.
+    _init_from_empy_active_files: bool,
+}
+
+impl<F: FileSystem> FileCollection<F> {
+    pub fn build(
+        file_system: Arc<F>,
+        queue: LogQueue,
+        cfg: &Config,
+        capacity: usize,
+        active_files: FileList<F>,
+        stale_files: FileList<F>,
+    ) -> Result<Self> {
+        let alignment = || {
+            fail_point!("file_mgr::file_collection::force_set_alignment", |_| { 16 });
+            0
+        };
+        let empty_active_files = active_files.len() == 0;
+        let file_collection = Self {
+            file_system,
+            queue,
+            paths: [Path::new(&cfg.dir).to_path_buf()],
+            file_format: LogFileFormat::new(cfg.format_version, alignment()),
+            target_file_size: cfg.target_file_size.0 as usize,
+            capacity,
+            active_files: CachePadded::new(RwLock::new(active_files)),
+            stale_files: CachePadded::new(RwLock::new(stale_files)),
+            _init_from_empy_active_files: empty_active_files,
+        };
+        file_collection.initialize(cfg.prefill_for_recycle)?;
+        Ok(file_collection)
+    }
+
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.active_files.read().len() + self.stale_files.read().len()
+    }
+
+    #[inline]
+    pub fn get_fd(&self, file_seq: FileSeq) -> Option<Arc<F::Handle>> {
+        self.active_files.read().get(file_seq)
+    }
+
+    #[inline]
+    pub fn active_file_span(&self) -> Option<(FileSeq, FileSeq)> {
+        self.active_files.read().span()
+    }
+
+    #[inline]
+    #[cfg(test)]
+    fn stale_file_span(&self) -> Option<(FileSeq, FileSeq)> {
+        self.stale_files.read().span()
+    }
+
+    #[inline]
+    #[cfg(test)]
+    pub fn push_back(&self, seq: FileSeq, handle: FileWithFormat<F>) -> FileState {
+        self.active_files.write().push_back(seq, handle)
+    }
+
+    #[inline]
+    pub fn target_file_size(&self) -> usize {
+        self.target_file_size
+    }
+
+    #[inline]
+    /// Returns the last active file handle to the caller.
+    ///
+    /// Attention, this func can only be called when the caller firstly
+    /// tried to access the latest active file.
+    pub fn fetch_active_file(&self) -> Option<ActiveFile<F>> {
```

Comment:
Remove the `_init_from_empy_active_files`, and change this function into `fn get_or_create_active_file(&mut self) -> (FileWithFormat, bool)`